File: internal/scanner/linux.go
//go:build linux

package scanner

import (
	"bufio"
	"context"
	"errors"
	"fmt"
	"github.com/karimz1/open-file-lock-handle/internal/model"
	"golang.org/x/sys/unix"
	"os"
	"os/user"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"syscall"
)

type native struct {
	root     string
	sampleMu sync.Mutex
	samples  map[string]cpuSample
}

func New() (Scanner, error) { return &native{root: "/proc"}, nil }

func (s *native) identity(pid int) (model.Identity, error) {
	b, err := os.ReadFile(filepath.Join(s.root, strconv.Itoa(pid), "stat"))
	if err != nil {
		return model.Identity{}, err
	}
	// comm is parenthesized and may itself contain spaces or closing parentheses.
	end := strings.LastIndexByte(string(b), ')')
	if end < 0 {
		return model.Identity{}, errors.New("invalid proc stat")
	}
	f := strings.Fields(string(b[end+1:]))
	if len(f) < 20 {
		return model.Identity{}, errors.New("short proc stat")
	}
	return model.Identity{PID: pid, Started: f[19]}, nil
}

func (s *native) Scan(ctx context.Context, target model.Target) (model.Result, error) {
	result := model.Result{LockDetection: true}
	entries, err := os.ReadDir(s.root)
	if err != nil {
		return result, fmt.Errorf("read proc: %w", err)
	}
	denied, foreign := 0, 0
	selfNS, _ := os.Readlink(filepath.Join(s.root, "self/ns/mnt"))
	users := make(map[string]string)
	for _, e := range entries {
		if err := ctx.Err(); err != nil {
			return result, err
		}
		pid, err := strconv.Atoi(e.Name())
		if err != nil || pid == os.Getpid() {
			continue
		}
		id, err := s.identity(pid)
		if err != nil {
			if errors.Is(err, os.ErrPermission) {
				denied++
			}
			continue
		}
		base := filepath.Join(s.root, e.Name())
		ns, _ := os.Readlink(filepath.Join(base, "ns/mnt"))
		if selfNS != "" && ns != "" && selfNS != ns {
			foreign++
			continue
		}
		p := model.Process{Identity: id, User: "unknown"}
		restricted := false
		check := func(err error) {
			if errors.Is(err, os.ErrPermission) {
				restricted = true
			}
		}
		status, err := os.ReadFile(filepath.Join(base, "status"))
		check(err)
		for _, line := range strings.Split(string(status), "\n") {
			if strings.HasPrefix(line, "Name:") {
				p.Name = strings.TrimSpace(strings.TrimPrefix(line, "Name:"))
			}
			if strings.HasPrefix(line, "Uid:") {
				f := strings.Fields(line)
				if len(f) > 2 {
					uid := f[2]
					name, ok := users[uid]
					if !ok {
						name = uid
						if u, e := user.LookupId(uid); e == nil {
							name = u.Username
						}
						users[uid] = name
					}
					p.User = name
				}
			}
		}
		add := func(path, relation, access, ref string) {
			deleted := strings.HasSuffix(path, " (deleted)")
			// /proc's deleted suffix is ambiguous for a live file literally named that.
			if deleted {
				if _, err := os.Stat(filepath.Join(base, "root", path)); err != nil {
					path = strings.TrimSuffix(path, " (deleted)")
				} else {
					deleted = false
				}
			}
			if !filepath.IsAbs(path) {
				return
			}
			var info os.FileInfo
			if ref != "" {
				info, _ = os.Stat(ref)
			}
			if target.Matches(path, info) {
				p.Usages = append(p.Usages, model.Usage{Path: path, Relation: relation, Access: access, Deleted: deleted})
			}
		}
		for _, rel := range []struct{ name, relation, access string }{{"cwd", "cwd", "directory"}, {"exe", "executable", "execute"}} {
			ref := filepath.Join(base, rel.name)
			path, err := os.Readlink(ref)
			check(err)
			if err == nil {
				if rel.name == "cwd" {
					p.CWD = path
				} else {
					p.Executable = path
				}
				add(path, rel.relation, rel.access, ref)
			}
		}
		fds, err := os.ReadDir(filepath.Join(base, "fd"))
		check(err)
		for _, fd := range fds {
			if err := ctx.Err(); err != nil {
				return result, err
			}
			ref := filepath.Join(base, "fd", fd.Name())
			path, err := os.Readlink(ref)
			check(err)
			if err != nil {
				continue
			}
			access := "unknown"
			b, err := os.ReadFile(filepath.Join(base, "fdinfo", fd.Name()))
			check(err)
			for _, line := range strings.Split(string(b), "\n") {
				if strings.HasPrefix(line, "flags:") {
					flags, e := strconv.ParseUint(strings.TrimSpace(strings.TrimPrefix(line, "flags:")), 8, 64)
					if e == nil {
						if flags&unix.O_PATH != 0 {
							access = "reference"
						} else {
							switch flags & unix.O_ACCMODE {
							case unix.O_RDONLY:
								access = "read"
							case unix.O_WRONLY:
								access = "write"
							case unix.O_RDWR:
								access = "read/write"
							}
						}
					}
				}
			}
			add(path, "open", access, ref)
			for _, lock := range fdLocks(string(b)) {
				before := len(p.Usages)
				add(path, "locked", access, ref)
				if len(p.Usages) > before {
					p.Usages[len(p.Usages)-1].Lock = lock
				}
			}
		}
		maps, err := os.Open(filepath.Join(base, "maps"))
		check(err)
		if err == nil {
			lines := bufio.NewScanner(maps)
			lines.Buffer(make([]byte, 4096), 1024*1024)
			for lines.Scan() {
				if err := ctx.Err(); err != nil {
					maps.Close()
					return result, err
				}
				line := lines.Text()
				f := strings.Fields(line)
				if len(f) < 6 {
					continue
				}
				// Preserve spaces in filenames; maps escapes embedded newlines as \\012.
				at := 0
				for i := 0; i < 5; i++ {
					for at < len(line) && line[at] == ' ' {
						at++
					}
					for at < len(line) && line[at] != ' ' {
						at++
					}
				}
				for at < len(line) && line[at] == ' ' {
					at++
				}
				path := strings.ReplaceAll(line[at:], `\012`, "\n")
				if !strings.HasPrefix(path, "/") {
					continue
				}
				if !target.Directory && target.Info != nil {
					if stat, ok := target.Info.Sys().(*syscall.Stat_t); ok {
						var major, minor uint32
						fmt.Sscanf(f[3], "%x:%x", &major, &minor)
						ino, _ := strconv.ParseUint(f[4], 10, 64)
						if ino != stat.Ino || major != unix.Major(uint64(stat.Dev)) || minor != unix.Minor(uint64(stat.Dev)) {
							continue
						}
					}
				}
				access := "mapped"
				if strings.Contains(f[1], "x") {
					access = "execute"
				} else if strings.HasPrefix(f[1], "rw") {
					access = "read/write"
				} else if strings.HasPrefix(f[1], "r") {
					access = "read"
				} else if len(f[1]) > 1 && f[1][1] == 'w' {
					access = "write"
				}
				add(path, "mapped", access, "")
			}
			check(lines.Err())
			maps.Close()
		}
		if restricted {
			denied++
		}
		if len(p.Usages) > 0 {
			current, e := s.identity(pid)
			if e == nil && current == id {
				result.Processes = append(result.Processes, p)
			}
		}
	}
	if denied > 0 {
		result.Warnings = append(result.Warnings, fmt.Sprintf("Limited visibility for %d processes (permissions). Elevated access may reveal more.", denied))
	}
	if foreign > 0 {
		result.Warnings = append(result.Warnings, fmt.Sprintf("Skipped %d processes in other mount namespaces; run oflh inside their container.", foreign))
	}
	s.enrich(ctx, &result)
	result.Normalize()
	return result, nil
}

func (s *native) Kill(ctx context.Context, id model.Identity, force bool) error {
	if err := validate(id); err != nil {
		return err
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	// A pidfd pins the process throughout validation and signal delivery.
	fd, err := unix.PidfdOpen(id.PID, 0)
	if err != nil {
		return fmt.Errorf("open process safely (requires Linux 5.3+): %w", err)
	}
	defer unix.Close(fd)
	current, err := s.identity(id.PID)
	if err != nil || current != id {
		return ErrChanged
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	signal := unix.SIGTERM
	if force {
		signal = unix.SIGKILL
	}
	if err := unix.PidfdSendSignal(fd, signal, nil, 0); err != nil {
		return fmt.Errorf("signal PID %d: %w", id.PID, err)
	}
	return nil
}

// fdLocks accepts held kernel locks only, excluding blocked requests and leases.
func fdLocks(info string) []string {
	var locks []string
	for _, line := range strings.Split(info, "\n") {
		f := strings.Fields(line)
		if len(f) != 9 || f[0] != "lock:" {
			continue
		}
		if f[2] != "FLOCK" && f[2] != "POSIX" && f[2] != "OFDLCK" {
			continue
		}
		if f[4] != "READ" && f[4] != "WRITE" {
			continue
		}
		locks = append(locks, fmt.Sprintf("%s %s %s bytes %s–%s", f[2], f[3], f[4], f[7], f[8]))
	}
	return locks
}

File: crates/oflh-platform/src/linux.rs
use super::*;
use rustix::process::{Pid, PidfdFlags, Signal, pidfd_open, pidfd_send_signal};
use std::{
    fs::{self, File},
    io::{BufRead, BufReader},
    os::unix::{ffi::OsStrExt, fs::MetadataExt},
    path::{Path, PathBuf},
};
#[derive(Default)]
pub struct Native {
    sampler: Sampler,
}
#[derive(Debug)]
struct Stat {
    identity: Identity,
    name: String,
    parent: u32,
    ticks: u64,
    rss: Option<u64>,
}
fn stat(pid: u32) -> Result<Stat> {
    let bytes = fs::read(format!("/proc/{pid}/stat"))
        .map_err(|error| io("read process identity", error))?;
    parse_stat(pid, &bytes).ok_or(Error::Changed)
}
fn parse_stat(pid: u32, bytes: &[u8]) -> Option<Stat> {
    let start = bytes.iter().position(|bytes| *bytes == b'(')?;
    let end = bytes.iter().rposition(|bytes| *bytes == b')')?;
    if end <= start {
        return None;
    }
    let mut field_iter = bytes[end + 1..]
        .split(|bytes| bytes.is_ascii_whitespace())
        .filter(|fields| !fields.is_empty());
    let mut fields = [&b""[..]; 22];
    for field in &mut fields {
        *field = field_iter.next()?
    }
    let parse_number = |i| std::str::from_utf8(fields[i]).ok()?.parse::<u64>().ok();
    // SAFETY: sysconf takes a constant selector and no pointers.
    let page = unsafe { libc::sysconf(libc::_SC_PAGESIZE) };
    Some(Stat {
        identity: Identity {
            pid,
            started: parse_number(19)?,
            started_sub: 0,
        },
        name: String::from_utf8_lossy(&bytes[start + 1..end]).into_owned(),
        parent: parse_number(1)?.try_into().ok()?,
        ticks: parse_number(11)?.checked_add(parse_number(12)?)?,
        rss: parse_number(21)
            .and_then(|parse_number| parse_number.checked_mul(page.try_into().ok()?)),
    })
}
fn total_cpu() -> u64 {
    fs::read_to_string("/proc/stat")
        .ok()
        .and_then(|s| {
            let mut fields = s.lines().next()?.split_whitespace();
            if fields.next() != Some("cpu") {
                return None;
            }
            fields.take(8).try_fold(0u64, |sum, parse_number| {
                sum.checked_add(parse_number.parse().ok()?)
            })
        })
        .unwrap_or(0)
}
fn deleted_path(base: &Path, mut path: PathBuf) -> (PathBuf, bool) {
    let raw = path.as_os_str().as_bytes();
    if let Some(stripped) = raw.strip_suffix(b" (deleted)")
        && fs::metadata(
            base.join("root")
                .join(path.strip_prefix("/").unwrap_or(&path)),
        )
        .is_err()
    {
        path = PathBuf::from(std::ffi::OsStr::from_bytes(stripped));
        return (path, true);
    }
    (path, false)
}
fn permission(error: &std::io::Error, restricted: &mut bool) {
    if error.kind() == std::io::ErrorKind::PermissionDenied {
        *restricted = true
    }
}
impl Backend for Native {
    fn scan(&mut self, target: &Target, cancel: &Cancellation) -> Result<Snapshot> {
        let mut snapshot = Snapshot::default();
        let mut denied = 0;
        let mut foreign = 0;
        let mut users = HashMap::new();
        let own_ns = fs::read_link("/proc/self/ns/mnt").ok();
        for entry in fs::read_dir("/proc").map_err(|error| io("enumerate processes", error))? {
            cancel.check()?;
            let Ok(entry) = entry else { continue };
            let Some(pid) = entry
                .file_name()
                .to_str()
                .and_then(|s| s.parse::<u32>().ok())
            else {
                continue;
            };
            if pid == std::process::id() {
                continue;
            }
            let stats = match stat(pid) {
                Ok(stats) => stats,
                Err(Error::Io { source, .. }) => {
                    if source.kind() == std::io::ErrorKind::PermissionDenied {
                        denied += 1
                    }
                    continue;
                }
                Err(_) => continue,
            };
            let base = entry.path();
            if let (Some(a), Ok(bytes)) = (&own_ns, fs::read_link(base.join("ns/mnt")))
                && *a != bytes
            {
                foreign += 1;
                continue;
            }
            let mut restricted = false;
            let mut process = Process {
                identity: stats.identity,
                name: stats.name,
                parent: stats.parent,
                memory: stats.rss,
                ..Process::default()
            };
            for (name, relation, access) in [
                ("cwd", Relation::Cwd, Access::Directory),
                ("exe", Relation::Executable, Access::Execute),
            ] {
                let reference = base.join(name);
                match fs::read_link(&reference) {
                    Ok(path) => {
                        if name == "cwd" {
                            process.cwd = path.clone()
                        } else {
                            process.executable = path.clone()
                        }
                        let (path, deleted) = deleted_path(&base, path);
                        if target.directory && !target.contains(&path) {
                            continue;
                        }
                        let metadata = if target.directory {
                            None
                        } else {
                            fs::metadata(reference).ok()
                        };
                        if target.matches(&path, metadata.as_ref()) {
                            process.usages.push(Usage {
                                path,
                                relation,
                                access,
                                deleted,
                                lock: None,
                            })
                        }
                    }
                    Err(error) => permission(&error, &mut restricted),
                }
            }
            match fs::read_dir(base.join("fd")) {
                Err(error) => permission(&error, &mut restricted),
                Ok(fds) => {
                    for descriptor in fds {
                        cancel.check()?;
                        let Ok(descriptor) = descriptor else { continue };
                        let reference = descriptor.path();
                        let path = match fs::read_link(&reference) {
                            Ok(process) => process,
                            Err(error) => {
                                permission(&error, &mut restricted);
                                continue;
                            }
                        };
                        if !path.is_absolute() {
                            continue;
                        }
                        let (path, deleted) = deleted_path(&base, path);
                        if target.directory && !target.contains(&path) {
                            continue;
                        }
                        let metadata = if target.directory {
                            None
                        } else {
                            fs::metadata(&reference).ok()
                        };
                        if !target.matches(&path, metadata.as_ref()) {
                            continue;
                        }
                        let descriptor_info = match fs::read_to_string(
                            base.join("fdinfo").join(descriptor.file_name()),
                        ) {
                            Ok(s) => s,
                            Err(error) => {
                                permission(&error, &mut restricted);
                                String::new()
                            }
                        };
                        let access = descriptor_info
                            .lines()
                            .find_map(|line| line.strip_prefix("flags:"))
                            .and_then(|s| u32::from_str_radix(s.trim(), 8).ok())
                            .map(|flags| {
                                if flags & libc::O_PATH as u32 != 0 {
                                    Access::Reference
                                } else {
                                    match flags & libc::O_ACCMODE as u32 {
                                        0 => Access::Read,
                                        1 => Access::Write,
                                        2 => Access::ReadWrite,
                                        _ => Access::Unknown,
                                    }
                                }
                            })
                            .unwrap_or_default();
                        process.usages.push(Usage {
                            path: path.clone(),
                            relation: Relation::Open,
                            access,
                            deleted,
                            lock: None,
                        });
                        for lock in locks(&descriptor_info) {
                            process.usages.push(Usage {
                                path: path.clone(),
                                relation: Relation::Locked,
                                access,
                                deleted,
                                lock: Some(LockEvidence::Kernel(lock)),
                            })
                        }
                    }
                }
            }
            match File::open(base.join("maps")) {
                Err(error) => permission(&error, &mut restricted),
                Ok(file) => {
                    let mut reader = BufReader::with_capacity(8192, file);
                    let mut line = Vec::with_capacity(512);
                    loop {
                        cancel.check()?;
                        line.clear();
                        match reader.read_until(b'\n', &mut line) {
                            Ok(0) => break,
                            Ok(_) => {}
                            Err(error) => {
                                permission(&error, &mut restricted);
                                break;
                            }
                        }
                        if let Some((path, access, dev, ino)) = mapping(&line) {
                            let (path, deleted) = deleted_path(&base, path);
                            if target.directory {
                                if !target.contains(&path) {
                                    continue;
                                }
                            } else if let Some(m) = &target.metadata {
                                if m.ino() != ino || m.dev() != dev {
                                    continue;
                                }
                            } else if !target.matches(&path, None) {
                                continue;
                            }
                            process.usages.push(Usage {
                                path,
                                relation: Relation::Mapped,
                                access,
                                deleted,
                                lock: None,
                            });
                        }
                    }
                }
            }
            if restricted {
                denied += 1
            }
            if !process.usages.is_empty() && stat(pid).is_ok_and(|s| s.identity == process.identity)
            {
                let uid = fs::read_to_string(base.join("status")).ok().and_then(|s| {
                    s.lines()
                        .find_map(|l| l.strip_prefix("Uid:"))
                        .and_then(|s| s.split_whitespace().nth(1))
                        .and_then(|s| s.parse::<u32>().ok())
                });
                process.user = uid
                    .map(|uid| {
                        users
                            .entry(uid)
                            .or_insert_with(|| super::unix::username(uid))
                            .clone()
                    })
                    .unwrap_or_else(|| "unknown".into());
                let mut next = process.parent;
                while next > 0
                    && process.ancestors.len() < 8
                    && next != pid
                    && !process.ancestors.iter().any(|a| a.identity.pid == next)
                {
                    cancel.check()?;
                    match stat(next) {
                        Ok(s) => {
                            process.ancestors.push(Ancestor {
                                identity: s.identity,
                                name: s.name,
                            });
                            next = s.parent
                        }
                        Err(_) => {
                            process.ancestors.push(Ancestor {
                                identity: Identity {
                                    pid: next,
                                    ..Identity::default()
                                },
                                name: "unavailable".into(),
                            });
                            break;
                        }
                    }
                }
                snapshot.processes.push(process);
            }
        }
        if denied > 0 {
            snapshot.warnings.push(format!("Limited visibility for {denied} processes (permissions). Elevated access may reveal more."))
        }
        if foreign > 0 {
            snapshot.warnings.push(format!("Skipped {foreign} processes in other mount namespaces; run oflh inside their container."))
        }
        snapshot.normalize();
        let ids = snapshot
            .processes
            .iter()
            .map(|process| process.identity)
            .collect::<Vec<_>>();
        apply_metrics(&mut snapshot, self.sample(&ids, cancel)?);
        Ok(snapshot)
    }
    fn sample(
        &mut self,
        ids: &[Identity],
        cancel: &Cancellation,
    ) -> Result<Vec<(Identity, Metrics)>> {
        let total = total_cpu();
        let mut raw = Vec::with_capacity(ids.len());
        for &identity in ids {
            cancel.check()?;
            if let Ok(stats) = stat(identity.pid)
                && stats.identity == identity
            {
                raw.push((identity, stats.ticks, stats.rss))
            }
        }
        if total == 0 {
            return Ok(raw
                .into_iter()
                .map(|(identity, _, memory)| (identity, Metrics { memory, cpu: None }))
                .collect());
        }
        Ok(self.sampler.sample(raw, total))
    }
    fn terminate(&mut self, identity: Identity, force: bool, cancel: &Cancellation) -> Result<()> {
        identity.validate()?;
        cancel.check()?;
        let pid = Pid::from_raw(identity.pid.try_into().map_err(|_| Error::Protected)?)
            .ok_or(Error::Protected)?;
        let descriptor = pidfd_open(pid, PidfdFlags::empty())
            .map_err(|error| io("open process safely (requires Linux 5.3+)", error.into()))?;
        if stat(identity.pid)?.identity != identity {
            return Err(Error::Changed);
        }
        cancel.check()?;
        pidfd_send_signal(&descriptor, if force { Signal::KILL } else { Signal::TERM })
            .map_err(|error| io("signal process", error.into()))
    }
}
fn locks(descriptor_info: &str) -> Vec<String> {
    descriptor_info
        .lines()
        .filter_map(|line| {
            let fields: Vec<_> = line.split_whitespace().collect();
            if fields.len() != 9
                || fields[0] != "lock:"
                || !matches!(fields[2], "FLOCK" | "POSIX" | "OFDLCK")
                || !matches!(fields[4], "READ" | "WRITE")
            {
                return None;
            }
            Some(format!(
                "{} {} {} bytes {}–{}",
                fields[2], fields[3], fields[4], fields[7], fields[8]
            ))
        })
        .collect()
}
fn mapping(line: &[u8]) -> Option<(PathBuf, Access, u64, u64)> {
    let mut at = 0;
    let mut fields = [&b""[..]; 5];
    for field in &mut fields {
        while line.get(at) == Some(&b' ') {
            at += 1
        }
        let start = at;
        while line.get(at).is_some_and(|bytes| *bytes != b' ') {
            at += 1
        }
        *field = line.get(start..at)?;
    }
    while line.get(at) == Some(&b' ') {
        at += 1
    }
    let path = line.get(at..)?.strip_suffix(b"\n").unwrap_or(&line[at..]);
    if !path.starts_with(b"/") {
        return None;
    }
    let mut decoded = Vec::with_capacity(path.len());
    let mut i = 0;
    while i < path.len() {
        if path[i..].starts_with(b"\\012") {
            decoded.push(b'\n');
            i += 4
        } else {
            decoded.push(path[i]);
            i += 1
        }
    }
    let dev = std::str::from_utf8(fields[3]).ok()?;
    let (major, minor) = dev.split_once(':')?;
    let dev = libc::makedev(
        u32::from_str_radix(major, 16).ok()?,
        u32::from_str_radix(minor, 16).ok()?,
    );
    let ino = std::str::from_utf8(fields[4]).ok()?.parse().ok()?;
    let perm = fields[1];
    let access = if perm.contains(&b'x') {
        Access::Execute
    } else if perm.starts_with(b"rw") {
        Access::ReadWrite
    } else if perm.starts_with(b"r") {
        Access::Read
    } else if perm.get(1) == Some(&b'w') {
        Access::Write
    } else {
        Access::Mapped
    };
    Some((
        PathBuf::from(std::ffi::OsStr::from_bytes(&decoded)),
        access,
        dev,
        ino,
    ))
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn held_locks_only() {
        assert_eq!(
            locks(
                "lock: 1: -> POSIX ADVISORY WRITE 42 00:13:12 0 EOF\nlock: 2: LEASE ACTIVE READ 42 00:13:12 0 EOF\nlock: 3: OFDLCK ADVISORY READ -1 00:13:12 4 8\n"
            ),
            ["OFDLCK ADVISORY READ bytes 4–8"]
        );
    }
    #[test]
    fn spaces_in_maps() {
        let (process, _, _, ino) =
            mapping(b"100-200 rw-p 0000 00:13 12  /a file\\012name\n").unwrap();
        assert_eq!(process, Path::new("/a file\nname"));
        assert_eq!(ino, 12);
    }
}
